packages/medusa/src/services/payment-provider.ts

```diff
@@ -55,6 +59,10 @@ export default class PaymentProviderService extends TransactionBaseService {
   // eslint-disable-next-line max-len
   protected readonly paymentProviderRepository_: typeof PaymentProviderRepository
   protected readonly paymentRepository_: typeof PaymentRepository
+  protected get paymentService_(): PaymentService {
+    // defer resolution. then it will use the cached resolved service
+    return this.container_.paymentService
```

Comment:
question: Services are registered as singletons in the container meaning they are already cached at the root. What's the difference? And why are we introducing this only for the PaymentService?

Reply:
Why can't we do: 
```
protected readonly paymentService_: PaymentService
...
this.paymentService_ = this.container_.paymentService
```

That would follow our general convention around service resolving

Reply:
Because we have a circular dependency, even though this is generally something to avoid, in some cases it is acceptable. In a class IoC approach, the resolution of a circular is to set one of the side to undefined and then assign the resolved value to it at a later point. Which is why in a classic framework using IoC it is recommended to not do anything in the constructor as an argument can be undefined for a moment.

Here I am just delaying the resolution of the dependency to a later point